--- net/bridge/br_input.c.orig	2014-08-04 00:25:02.000000000 +0200
+++ net/bridge/br_input.c	2015-04-13 14:35:27.524951138 +0200
@@ -202,7 +202,11 @@
 			break;
 
 		case 0x01:	/* IEEE MAC (Pause) */
-			goto drop;
+			fwd_mask |= p->br->group_fwd_mask;
+			if (fwd_mask & (1u << dest[5]))
+				goto forward;
+            else
+				goto drop;
 
 		default:
 			/* Allow selective forwarding for most other protocols */
--- net/bridge/br_private.h.orig	2015-04-13 14:10:25.000000000 +0200
+++ net/bridge/br_private.h	2015-04-13 14:36:09.540949915 +0200
@@ -33,8 +33,6 @@
 
 /* Control of forwarding link local multicast */
 #define BR_GROUPFWD_DEFAULT	0
-/* Don't allow forwarding control protocols like STP and LLDP */
-#define BR_GROUPFWD_RESTRICTED	0x4007u
 /* The Nearest Customer Bridge Group Address, 01-80-C2-00-00-[00,0B,0C,0D,0F] */
 #define BR_GROUPFWD_8021AD	0xB801u
 
--- net/bridge/br_sysfs_br.c.orig	2014-08-04 00:25:02.000000000 +0200
+++ net/bridge/br_sysfs_br.c	2015-04-13 14:36:24.960949466 +0200
@@ -171,9 +171,6 @@
 	if (endp == buf)
 		return -EINVAL;
 
-	if (val & BR_GROUPFWD_RESTRICTED)
-		return -EINVAL;
-
 	br->group_fwd_mask = val;
 
 	return len;
